README.md
# gochariots
Scalable shared log *Chariots* implemented in Go.
For the system details, please refer to the EDBT paper:
[Chariots : A Scalable Shared Log for Data Management in Multi-Datacenter Cloud Environments](https://openproceedings.org/2015/conf/edbt/paper-236.pdf)

## Using Docker
If you are seeing this page on Docker repository page, [please see this page for details](https://github.com/fasthall/gochariots/blob/master/doc/docker.md).

## Quickstart
We can use Docker Compose to run a local shared log, or Docker Swarm to run a cluster on multiple hosts. Please see [this document](doc/docker_stack.md) for detail.

## Appending record
To append to the shared log, send POST request to http://localhost:8080/record. The payload needs to be in JSON format. See [post_example.sh](test/post_example.sh) and [example.json](test/example.json).

## Issues (TODO)
When a component of any stage stops working and the controller couldn't connect to it, the controller should prompt a warning message rather than just panic. 

## Design discussion

* Currently sender(propogation) is bundled with maintainer. This eliminate the communication between sender and maintainer, but can't be scaled. If seperate them, sender can be scaled independently, but require extra communication.
* Does token need to carry all records with unsatisfied dependencies? Currently it does that.

doc/changelog.md
# October

## October 17, 2017
### Two-phase append
When a record hits queue, the queue immediately appends it to the maintainer without LogID assigned. The record's UUID and its parent's UUID is buffered in the queue. When the parent's UUID presents in the maintainers, the queue assign LogID and update the entries in maintainer.

## October 5, 2017
### Change datacenter ID numbering
Now dc ID in config.yaml starts from 1. When create a new record using library, host id will be 0. If a batcher receive a record with host id 0, it will fill in host id as dc id automatically.

## October 3, 2017
### Replace comunication code with grpc
To improve reliability and performance

# August

## August 24, 2017
### Add get() function to TOId version
When posting a record to TOId version of GoChariots, a unique ID needs to be assigned. After that, user can send a GET request to app to query the record.

## August 23, 2017
### Remove filter
In the current design there is no need of filter.

## August 14, 2017
### Add Azure Cosmos DB as data store option
User can use Azure Cosmos DB to store records now. 

## August 10, 2017
### Multiple prerequiste hashes support
Now an event record can have more than one prerequiste hash. For example, when record `r3` has hashes of both `r1` and `r2`, both `r1` and `r2` need to be appended before `r3` does. 

# July

## July 

## July 28, 2017
### Add Google Datastore adapter
User can use Datastore to store records now. [Detail](../doc/adapter_datastore.md)

## July 27, 2017
### Add DynamoDB adapter
User can use DynamoDB to store records now. [Detail](../doc/adapter_dynamodb.md)

## July 25, 2017
### Bug fixed
Fix a problem gochariots may fetch wrong local IP when run as docker container.

## July 16, 2017
### Launch option change
Give default value to num_dc and id. By default, when launching a new component, the number of DC is 1 and the id of the said component is 0. To change the value, use option *--num_dc** and *--id*. Run gochariots --help for details. 

## July 12, 2017
### Docker Swarm support
A [docker-stack.yaml](../deploy/docker-stack.yaml) file is provided for quick deploying a log over Docker Swarm network. See the [instructions](docker_stack.md).

## July 10, 2017
### Docker Compose support
A [docker-compose.yaml](../deploy/docker-compose.yaml) file is provided for quick deploying a log on localhost. See the [instructions](docker_stack.md).

# June

## June 29, 2017
### Bug fixed
Fix a bug that sometimes app doesn't receive update on batcher list. 

## June 28, 2017
### Bug fixed
Fix a bug that controller won't notify queues about indexer. This bug happened when indexer is ready before queue is reported.

### App report to controller
Now apps will report to controller too, if option *-f* is used and config file is provided.

## June 27, 2017
### Rewrite how component reports to controller
Use *--config_file* or *-f* option to specify the YAML file location. In this file, user can specify the controller's host address. If the option is used, the component will report to the controller automatically until succeed. [Example of yaml file](../test/script/config1.yaml). [Example of new launch script](../test/script/double.sh). 

## June 22, 2017
### Merge all three branches
Use *-toid* option to use TOId version. By default use hash version. See [double_toid.sh](../test/script/double_toid.sh) as reference.

## June 21, 2017
### Use BoltDB to store indexes in indexer
By default, all the indexes(hash and seed pair) are stored in memory. An instance with 4GB memory can roughly store 500 million entries. If exceeded, there's no overflow handling currently implemented. Using *-db* option to launch indexer will tell the indexer to use BoltDB and go-cache to access indexes. Each will be in the cache for 5 minutes then expire. The BoltDB file is stored as `indexes.db`. *Notice that this implementation is very slow, not optimized now*

## June 18, 2017
### Merge pretoid and no_deferred branches
There are only master and pretoid branches now. In master branch we use hash, while in pretoid branch we use toid. By default, token doesn't carry deferred records. To carry deferred records with token, use *-c* option, for example: `gochariots-queue -c 9020 1 0 true`.

## June 16, 2017
### Rewrite system logging using logrus
Rewrite system logging using [logrus](https://github.com/sirupsen/logrus), use *-v* option to turn on all logging infos. For example, `gochariots-app 8080 1 0` by default only write `WARNING` and `ERROR` level events into logging file. `gochariots-app -v 8080 1 0` will turn on all events of logging.

## June 8, 2017
### Compare the performance between carrying v.s. not carrying deferred records with token
Not carrying deferred records with token may cause additional delay, but it's negligible. See [0606.md](experiment/0606.md).

### Use hash and seed to describe causal dependency
A record now has Seed and Hash field. Seed is a randomly generated 64 bits number, or application can give it context and manually assign it to the record. All the events(records) in the same event chain will share an identical seed. Hash is used to specify causal dependency(see changelog of May 15), and serves as the key of indexer.

When a record is stored in maintainer, indexer will hash the content and use the hash as key to store record's seed as value. Now, the deferred record can query indexer whether its prerequisite has been indexed.

# May

## May 18, 2017
### Refactor connection reading function
As in [connection.go](../misc/connection/connection.go).

### Use hash to specify casual dependency instead of tags
Now, record has a field pre:hash which is a uint64 type. To specify casual dependency, firstly hash "key:value" into a 64 bits int using fnv-1a, then put this hash value into the field. The deferred record carried by token will be checked, if the hash value has been indexed by indexers, the queue will remove the record from the token and send it to maintainer. 


gochariots.go
package main

import (
	"fmt"
	"net"
	"os"

	"google.golang.org/grpc"
	"google.golang.org/grpc/reflection"

	"github.com/fasthall/gochariots/app"
	"github.com/fasthall/gochariots/batcher"
	"github.com/fasthall/gochariots/batcher/batcherrpc"
	"github.com/fasthall/gochariots/controller"
	"github.com/fasthall/gochariots/info"
	"github.com/fasthall/gochariots/maintainer"
	"github.com/fasthall/gochariots/maintainer/adapter"
	"github.com/fasthall/gochariots/queue"

	"github.com/Sirupsen/logrus"
	kingpin "gopkg.in/alecthomas/kingpin.v2"
)

const version string = "mongodb 1102"

var (
	gochariots = kingpin.New("gochariots", "A distributed shared log system.")

	versionCommand = gochariots.Command("version", "The version of GoChariots.")

	appCommand = gochariots.Command("app", "Start an app instance.")
	appNumDC   = appCommand.Flag("num_dc", "The port app listens to.").Int()
	appID      = appCommand.Flag("id", "The port app listens to.").Int()
	appPort    = appCommand.Flag("port", "The port app listens to. By default it's 8080").Short('p').String()
	appTOId    = appCommand.Flag("toid", "Use TOId version.").Short('t').Bool()
	appConfig  = appCommand.Flag("config_file", "Configuration file to read.").Short('f').String()
	appInfo    = appCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	appDebug   = appCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()

	batcherCommand    = gochariots.Command("batcher", "Start a batcher instance.")
	batcherNumDC      = batcherCommand.Flag("num_dc", "The port batcher listens to.").Int()
	batcherID         = batcherCommand.Flag("id", "The port batcher listens to.").Int()
	batcherBufferSize = batcherCommand.Flag("buffer_size", "The buffer size.").Default("2000").Int()
	batcherPort       = batcherCommand.Flag("port", "The port app listens to. By default it's 9000").Short('p').String()
	batcherTOId       = batcherCommand.Flag("toid", "Use TOId version.").Short('t').Bool()
	batcherConfig     = batcherCommand.Flag("config_file", "Configuration file to read.").Short('f').String()
	batcherInfo       = batcherCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	batcherDebug      = batcherCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()

	controllerCommand = gochariots.Command("controller", "Start a controller instance.")
	controllerNumDC   = controllerCommand.Flag("num_dc", "The port controller listens to.").Int()
	controllerID      = controllerCommand.Flag("id", "The port controller listens to.").Int()
	controllerPort    = controllerCommand.Flag("port", "The port app listens to. By default it's 8081").Short('p').String()
	controllerInfo    = controllerCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	controllerDebug   = controllerCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()

	filterCommand = gochariots.Command("filter", "Start a filter instance.")
	filterNumDC   = filterCommand.Flag("num_dc", "The port filter listens to.").Int()
	filterID      = filterCommand.Flag("id", "The port contrfilteroller listens to.").Int()
	filterPort    = filterCommand.Flag("port", "The port app listens to. By default it's 9010").Short('p').String()
	filterTOId    = filterCommand.Flag("toid", "Use TOId version.").Short('t').Bool()
	filterConfig  = filterCommand.Flag("config_file", "Configuration file to read.").Short('f').String()
	filterInfo    = filterCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	filterDebug   = filterCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()

	queueCommand        = gochariots.Command("queue", "Start a queue instance.")
	queueNumDC          = queueCommand.Flag("num_dc", "The port queue listens to.").Int()
	queueID             = queueCommand.Flag("id", "The port queue listens to.").Int()
	queuePort           = queueCommand.Flag("port", "The port app listens to. By default it's 9020").Short('p').String()
	queueHold           = queueCommand.Flag("hold", "Whether this queue instance holds a token when launched.").Required().Short('h').Bool()
	queueTOId           = queueCommand.Flag("toid", "Use TOId version.").Short('t').Bool()
	queueCarry          = queueCommand.Flag("carry", "Carry deferred records with token. Only work when toid is on.").Short('c').Bool()
	queueConfig         = queueCommand.Flag("config_file", "Configuration file to read.").Short('f').String()
	queueInfo           = queueCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	queueDebug          = queueCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()
	queueTwoPhaseAppend = queueCommand.Flag("two_phase_append", "Enable two phase append.").Bool()

	maintainerCommand   = gochariots.Command("maintainer", "Start a maintainer instance.")
	maintainerNumDC     = maintainerCommand.Flag("num_dc", "The port maintainer listens to.").Int()
	maintainerID        = maintainerCommand.Flag("id", "The port maintainer listens to.").Int()
	maintainerPort      = maintainerCommand.Flag("port", "The port app listens to. By default it's 9030").Short('p').String()
	maintainerTOId      = maintainerCommand.Flag("toid", "Use TOId version.").Short('t').Bool()
	maintainerConfig    = maintainerCommand.Flag("config_file", "Configuration file to read.").Short('f').String()
	maintainerInfo      = maintainerCommand.Flag("info", "Turn on info level logging.").Short('i').Bool()
	maintainerDebug     = maintainerCommand.Flag("debug", "Turn on debug level logging.").Short('d').Bool()
	maintainerDynamoDB  = maintainerCommand.Flag("dynamodb", "Use DynamoDB as physical storage.").Bool()
	maintainerDatastore = maintainerCommand.Flag("datastore", "Use Datastore as physical storage.").Bool()
	maintainerCosmosDB  = maintainerCommand.Flag("cosmosdb", "Use CosmosDB as physical storage.").Bool()
	maintainerMongoDB   = maintainerCommand.Flag("mongodb", "Use MongoDB as physical storage.").Bool()
)

func main() {
	switch kingpin.MustParse(gochariots.Parse(os.Args[1:])) {
	case versionCommand.FullCommand():
		fmt.Println(version)
	case appCommand.FullCommand():
		if *appPort == "" {
			*appPort = "8080"
		}
		if *appNumDC == 0 {
			*appNumDC = 1
		}
		info.NumDC = *appNumDC
		info.ID = *appID
		info.SetName("app" + *appPort)
		info.SetPort(*appPort)
		level := logrus.WarnLevel
		if *appDebug {
			level = logrus.DebugLevel
		} else if *appInfo {
			level = logrus.InfoLevel
		}
		info.RedirectLog(info.GetName()+".log", level)
		if *appConfig != "" {
			info.Config(*appConfig, "app")
		}
		if *appTOId {
			app.TOIDRun(*appPort)
		} else {
			app.Run(*appPort)
		}
	case controllerCommand.FullCommand():
		if *controllerPort == "" {
			*controllerPort = "8081"
		}
		if *controllerNumDC == 0 {
			*controllerNumDC = 1
		}
		info.NumDC = *controllerNumDC
		info.ID = *controllerID
		info.SetName("controller" + *controllerPort)
		info.SetPort(*controllerPort)
		level := logrus.WarnLevel
		if *controllerDebug {
			level = logrus.DebugLevel
		} else if *controllerInfo {
			level = logrus.InfoLevel
		}
		info.RedirectLog(info.GetName()+".log", level)
		controller.StartController(*controllerPort)
	case batcherCommand.FullCommand():
		if *batcherPort == "" {
			*batcherPort = "9000"
		}
		if *batcherNumDC == 0 {
			*batcherNumDC = 1
		}
		info.NumDC = *batcherNumDC
		info.ID = *batcherID
		info.SetName("batcher" + *batcherPort)
		info.SetPort(*batcherPort)
		level := logrus.WarnLevel
		if *batcherDebug {
			level = logrus.DebugLevel
		} else if *batcherInfo {
			level = logrus.InfoLevel
		}
		info.RedirectLog(info.GetName()+".log", level)
		if *batcherConfig != "" {
			info.Config(*batcherConfig, "batcher")
		}
		if *batcherTOId {
			batcher.TOIDInitBatcher(*batcherBufferSize)
		} else {
			batcher.InitBatcher(*batcherBufferSize)
		}
		ln, err := net.Listen("tcp", ":"+*batcherPort)
		if err != nil {
			panic(err)
		}
		defer ln.Close()
		fmt.Println(info.GetName()+" is listening to port", *batcherPort)
		s := grpc.NewServer()
		batcherrpc.RegisterBatcherRPCServer(s, &batcher.Server{})
		reflection.Register(s)
		if err := s.Serve(ln); err != nil {
			logrus.Fatalf("failed to serve: %v", err)
		}
	case queueCommand.FullCommand():
		if *queuePort == "" {
			*queuePort = "9020"
		}
		if *queueNumDC == 0 {
			*queueNumDC = 1
		}
		info.NumDC = *queueNumDC
		info.ID = *queueID
		info.SetName("queue" + *queuePort)
		info.SetPort(*queuePort)
		level := logrus.WarnLevel
		if *queueDebug {
			level = logrus.DebugLevel
		} else if *queueInfo {
			level = logrus.InfoLevel
		}
		info.RedirectLog(info.GetName()+".log", level)
		if *queueConfig != "" {
			info.Config(*queueConfig, "queue")
		}
		if *queueTOId {
			queue.TOIDInitQueue(*queueHold, *queueCarry)
		} else {
			queue.InitQueue(*queueHold, *queueTwoPhaseAppend)
		}
		ln, err := net.Listen("tcp", ":"+*queuePort)
		if err != nil {
			panic(err)
		}
		defer ln.Close()
		fmt.Println(info.GetName()+" is listening to port", *queuePort)
		s := grpc.NewServer()
		queue.RegisterQueueServer(s, &queue.Server{})
		reflection.Register(s)
		if err := s.Serve(ln); err != nil {
			logrus.Fatalf("failed to serve: %v", err)
		}
	case maintainerCommand.FullCommand():
		if *maintainerPort == "" {
			*maintainerPort = "9030"
		}
		if *maintainerNumDC == 0 {
			*maintainerNumDC = 1
		}
		info.NumDC = *maintainerNumDC
		info.ID = *maintainerID
		info.SetName("maintainer" + *maintainerPort)
		info.SetPort(*maintainerPort)
		level := logrus.WarnLevel
		if *maintainerDebug {
			level = logrus.DebugLevel
		} else if *maintainerInfo {
			level = logrus.InfoLevel
		}
		info.RedirectLog(info.GetName()+".log", level)
		if *maintainerConfig != "" {
			info.Config(*maintainerConfig, "maintainer")
		}
		adap := adapter.FLSTORE
		if *maintainerDynamoDB {
			adap = adapter.DYNAMODB
		} else if *maintainerDatastore {
			adap = adapter.DATASTORE
		} else if *maintainerCosmosDB {
			adap = adapter.COSMOSDB
		} else if *maintainerMongoDB {
			adap = adapter.MONGODB
		}
		maintainer.InitLogMaintainer(info.GetName(), adap)
		ln, err := net.Listen("tcp", ":"+*maintainerPort)
		if err != nil {
			fmt.Println(info.GetName() + "couldn't listen on port " + *maintainerPort)
			panic(err)
		}
		defer ln.Close()
		fmt.Println(info.GetName()+" is listening to port", *maintainerPort)
		s := grpc.NewServer()
		maintainer.RegisterMaintainerServer(s, &maintainer.Server{})
		reflection.Register(s)
		if err := s.Serve(ln); err != nil {
			logrus.Fatalf("failed to serve: %v", err)
		}
	}
}
